### PR TITLE
Fix a bug when using distributed tree with rays

### PR DESCRIPTION
**Pull request by Rombur**
There is a problem when using `Ray` and `DistributedTree` together; we don't define a `get` function when the `Geometry` is a  `Ray`. This PR fixes the bug and add a test. I've also added a call to `Details::check_valid_access_traits` to make sure that  the `Predicate` we create make sense. Without it the error message is really obscure (you get an error about `Kokkos::nonesuch`...)

### CODE REVIEW

File: src/details/ArborX_DetailsDistributedTreeImpl.hpp [hunk 1]
```diff
@@ -103,6 +104,15 @@ struct AccessTraits<
     auto const distance = x.distances(i);
     return intersects(Sphere{sphere.centroid(), distance + sphere.radius()});
   }
+  template <class Dummy = Geometry,
+            std::enable_if_t<std::is_same<Dummy, Geometry>::value &&
+                             std::is_same<Dummy, Experimental::Ray>::value> * =
+                nullptr>
```
Comment (aprokop) [inline src/details/ArborX_DetailsDistributedTreeImpl.hpp:110]: You could use `is_same_v` here.

File: test/tstDistributedTree.cpp [hunk 2]
```diff
@@ -813,3 +813,106 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(boost_comparison, DeviceType, ARBORX_DEVICE_TYPES)
...
+  }
+
+private:
+  Kokkos::View<ArborX::Experimental::Ray *, MemorySpace> _rays;
+};
+
+namespace ArborX
+{
+template <typename MemorySpace>
+struct AccessTraits<RayNearestPredicate<MemorySpace>, PredicatesTag>
```
Comment (aprokop) [inline test/tstDistributedTree.cpp:840]: Just use
```suggestion
struct ArborX::AccessTraits<RayNearestPredicate<MemorySpace>, PredicatesTag>
```
without namespace scope.

File: test/tstDistributedTree.cpp [hunk 3]
```diff
@@ -813,3 +813,106 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(boost_comparison, DeviceType, ARBORX_DEVICE_TYPES)
...
+                           make_reference_solution<PairIndexRank>(
+                               {{0, comm_rank}, {0, 0}}, {0, 1, 2}));
+  }
+  else
+  {
+    ARBORX_TEST_QUERY_TREE(ExecutionSpace{}, tree, RayNearestPredicate(rays),
+                           make_reference_solution<PairIndexRank>(
+                               {{0, comm_rank}, {0, comm_rank}}, {0, 1, 2}));
+  }
+}
```
Comment (aprokop) [inline test/tstDistributedTree.cpp:918]: Why did you need this `else`? It corresponds to `comm_rank = 0`, and is equivalent to the the code in `if` branch in that case.

**Review by masterleinad — APPROVED** [review]

**Review by aprokop — APPROVED** [review]

File: test/tstDistributedTree.cpp [hunk 4]
```diff
@@ -813,3 +814,95 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(boost_comparison, DeviceType, ARBORX_DEVICE_TYPES)
...
+          ArborX::Experimental::Vector direction{0.f, 1.f, 0.f};
+          rays(i) = ArborX::Experimental::Ray{origin, direction};
+        }
+        else
+        {
+          ArborX::Point origin{-0.5f, 0.5f, 0.5f};
+          ArborX::Experimental::Vector direction{1.f, 0.f, 0.f};
+          rays(i) = ArborX::Experimental::Ray{origin, direction};
+        }
+      });
```
Comment (dalg24) [inline test/tstDistributedTree.cpp:903]: This is a little hard to read as compared to something like
```C++
std::vector<ArborX::Experimental::Ray> rays = {
  {{comm_rank + 0.5f, -0.5f, 0.5f}, {0.f, 1.f, 0.f}},
  {{-0.5f, 0.5f, 0.5f}, {1.f, 0.f, 0.f}},
};


... RayNearestPredicate(toView<MemorySpace>(rays))

File: test/tstDistributedTree.cpp [hunk 5]
```diff
@@ -185,11 +186,12 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(empty_tree, DeviceType, ARBORX_DEVICE_TYPES)
...
-                           makeIntersectsSphereQueries<DeviceType>({
-                               {{{(float)comm_rank, 0.f, 0.f}}, (float)comm_size},
-                           }),
-                           make_reference_solution<PairIndexRank>({}, {0, 0}));
+    ARBORX_TEST_QUERY_TREE(
+        ExecutionSpace{}, tree,
+        makeIntersectsSphereQueries<DeviceType>({
+            {{{(float)comm_rank, 0.f, 0.f}}, (float)comm_size},
+        }),
+        make_reference_solution<PairIndexRank>({}, {0, 0}));
```
Comment (dalg24) [inline test/tstDistributedTree.cpp:194]: Are these the changes from #787 ?

**Review by dalg24 — APPROVED** [review]
